File: ros2_fundamentals/actions/action_tutorial/src/action_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "custom_msg_and_srv/action/concatenate.hpp"

class ActionClient : public rclcpp::Node
{
public:
    using Concatenate = custom_msg_and_srv::action::Concatenate;
    using GoalHandle  = rclcpp_action::ClientGoalHandle<Concatenate>;

    ActionClient() : Node("action_client")
    {
        // ----------------------------------------------------
        // CRÉER LE CLIENT D'ACTION
        // ----------------------------------------------------
        client_ = rclcpp_action::create_client<Concatenate>(
            this,
            "concatenate"   // même nom que le server
        );

        // Attendre que le server soit disponible
        while (!client_->wait_for_action_server(std::chrono::seconds(1)))
        {
            RCLCPP_WARN(this->get_logger(), "Waiting for action server...");
        }

        RCLCPP_INFO(this->get_logger(), "Action server found ! Sending goal...");

        // ----------------------------------------------------
        // CRÉER ET ENVOYER LE GOAL
        // ----------------------------------------------------
        auto goal = Concatenate::Goal();
        goal.first_string  = "Hello";
        goal.second_string = "ROS2 Actions !";

        // Options : définir les callbacks
        auto options = rclcpp_action::Client<Concatenate>::SendGoalOptions();

        // Callback quand le goal est accepté/rejeté
        options.goal_response_callback =
            std::bind(&ActionClient::goalResponseCallback, this,
                std::placeholders::_1);

        // Callback quand on reçoit un feedback
        options.feedback_callback =
            std::bind(&ActionClient::feedbackCallback, this,
                std::placeholders::_1, std::placeholders::_2);

        // Callback quand le goal est terminé
        options.result_callback =
            std::bind(&ActionClient::resultCallback, this,
                std::placeholders::_1);

        client_->async_send_goal(goal, options);
    }

private:
    rclcpp_action::Client<Concatenate>::SharedPtr client_;

    // --------------------------------------------------------
    // CALLBACK 1 : réponse du server (accepté/rejeté)
    // --------------------------------------------------------
    void goalResponseCallback(const GoalHandle::SharedPtr & goal_handle)
    {
        if (!goal_handle)
            RCLCPP_ERROR(this->get_logger(), "Goal rejected !");
        else
            RCLCPP_INFO(this->get_logger(), "Goal accepted !");
    }

    // --------------------------------------------------------
    // CALLBACK 2 : feedback reçu pendant l'exécution
    // --------------------------------------------------------
    void feedbackCallback(
        GoalHandle::SharedPtr,
        const std::shared_ptr<const Concatenate::Feedback> feedback)
    {
        RCLCPP_INFO(this->get_logger(), "Feedback: %s",
            feedback->current_status.c_str());
    }

    // --------------------------------------------------------
    // CALLBACK 3 : résultat final
    // --------------------------------------------------------
    void resultCallback(const GoalHandle::WrappedResult & result)
    {
        switch (result.code)
        {
            case rclcpp_action::ResultCode::SUCCEEDED:
                RCLCPP_INFO(this->get_logger(), "Result: '%s'",
                    result.result->concatenated_string.c_str());
                break;
            case rclcpp_action::ResultCode::CANCELED:
                RCLCPP_WARN(this->get_logger(), "Goal was canceled !");
                break;
            case rclcpp_action::ResultCode::ABORTED:
                RCLCPP_ERROR(this->get_logger(), "Goal was aborted !");
                break;
            default:
                break;
        }
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<ActionClient>());
    rclcpp::shutdown();
    return 0;
}

File: ros2_fundamentals/actions/action_tutorial_py/action_tutorial_py/action_client.py
import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from custom_msg_and_srv.action import Concatenate

class ConcatenateActionClient(Node):
    def __init__(self):
        super().__init__("action_client")

        # ----------------------------------------------------
        # CRÉER LE CLIENT D'ACTION
        # ----------------------------------------------------
        self.client_ = ActionClient(
            self,
            Concatenate,     # type de l'action
            "concatenate"    # nom de l'action
        )

        # Attendre que le server soit disponible
        while not self.client_.wait_for_server(timeout_sec=1.0):
            self.get_logger().warn("Waiting for action server...")

        self.get_logger().info("Action server found ! Sending goal...")

        # ----------------------------------------------------
        # CRÉER ET ENVOYER LE GOAL
        # ----------------------------------------------------
        goal = Concatenate.Goal()
        goal.first_string  = "Hello"
        goal.second_string = "ROS2 Actions !"

        # Envoyer le goal avec les callbacks
        self.client_.send_goal_async(
            goal,
            feedback_callback=self.feedback_callback
        ).add_done_callback(self.goal_response_callback)

    # --------------------------------------------------------
    # CALLBACK 1 : réponse du server
    # --------------------------------------------------------
    def goal_response_callback(self, future):
        goal_handle = future.result()

        if not goal_handle.accepted:
            self.get_logger().error("Goal rejected !")
            return

        self.get_logger().info("Goal accepted !")

        # Attendre le résultat final
        goal_handle.get_result_async().add_done_callback(
            self.result_callback
        )

    # --------------------------------------------------------
    # CALLBACK 2 : feedback pendant l'exécution
    # --------------------------------------------------------
    def feedback_callback(self, feedback_msg):
        self.get_logger().info(
            f"Feedback: {feedback_msg.feedback.current_status}"
        )

    # --------------------------------------------------------
    # CALLBACK 3 : résultat final
    # --------------------------------------------------------
    def result_callback(self, future):
        result = future.result().result
        self.get_logger().info(f"Result: '{result.concatenated_string}'")

def main(args=None):
    rclpy.init(args=args)
    node = ConcatenateActionClient()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()